### PR TITLE
Idris2: introduce `idris2Packages`, `idris2api`, and the `buildIdris` helper.

### PR DESCRIPTION
**Pull request by mattpolzin**
## Description of changes

The git diff isn't terribly helpful in identifying the fact that `default.nix` was renamed to `idris2.nix` with no substantive changes (I corrected a typo in one comment).

The new `default.nix` houses the `idris2Packages` namespace. `idris2` is still exposed at the top level, but it now also lives within `idris2Packages`.

I've also added the `buildIdris` helper which comes from the upstream Idris2 repository originally, though the same helper is being updated upstream at the same time as I am pushing improvements to it here in `nixpkgs` so the two versions will be a bit different until I've finished up everywhere (the end goal is rough parity).

This PR also adds `idris2Packages.idris2Api` which is an API offered by the compiler project; adding `idris2Api` to `nixpkgs` is a stepping stone on the way to adding the `idris2-lsp` language server to `nixpkgs` in a future PR.

Finally, I added a new documentation section for Idris 2 that mostly focuses on `buildIdris` for now.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: doc/languages-frameworks/idris2.section.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+```
+
+The above results in a derivation with the installed library results (with sourcecode).
+
+A slightly more involved example of a fully packaged executable would be the [`idris2-lsp`](https://github.com/idris-community/idris2-lsp) which is an Idris2 language server that uses the `LSP-lib` found above.
+```nix
+{ stdenv, fetchFromGitHub, idris2Packages }:
+
+# Assuming the previous example lives in `lsp-lib.nix`:
+let lspLib = stdenv.callPackage ./lsp-lib.nix { };
```
Comment (mattpolzin) [inline doc/languages-frameworks/idris2.section.md:32]:
```suggestion
{ callPackage, fetchFromGitHub, idris2Packages }:

# Assuming the previous example lives in `lsp-lib.nix`:
let lspLib = callPackage ./lsp-lib.nix { };
```

**Comment by mattpolzin** [comment]
Result of `nixpkgs-review pr 278391` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>idris2Packages.idris2</li>
  </ul>
</details>

**Comment by mattpolzin** [comment]
Result of `nixpkgs-review pr 278391` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>idris2Packages.idris2</li>
  </ul>
</details>

**Comment by mattpolzin** [comment]
@fabianhjr @wchresta let me know how this looks when you get the chance. I roughly followed the lead of the Elm packages structurally because it seemed simplest. 

I've got plans to put up a PR that adds the Idris2 LSP server once there's a foundation in place.

**Review by wchresta — APPROVED** [review]
Looks great! Thanks for doing this!

**Review by fabianhjr — APPROVED** [review]
lgtm, apologies forgot to follow up due to new year.

**Comment by mattpolzin** [comment]
Thank you both!

File: pkgs/development/compilers/idris2/idris2.nix [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  '';
+
+  # Run package tests
+  passthru.tests = callPackage ./tests.nix { inherit pname; };
+
+  meta = {
+    description = "A purely functional programming language with first class types";
+    homepage = "https://github.com/idris-lang/Idris2";
+    license = lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ fabianhjr wchresta ];
```
Comment (fabianhjr) [inline pkgs/development/compilers/idris2/idris2.nix:94]: Would you like to be a maintainer btw?

It is mostly about reviewing future proposed changes :3

Comment (mattpolzin) [reply pkgs/development/compilers/idris2/idris2.nix:94]: I'd be honored. I'll add a commit momentarily.

Comment (mattpolzin) [reply pkgs/development/compilers/idris2/idris2.nix:94]: Actually, would it be alright to ask you to merge an unrelated open PR that adds me to the maintainers file itself prior to me then adding myself to the list of maintainers on this PR? The other PR is approved-by-maintainers. https://github.com/NixOS/nixpkgs/pull/280825

Comment (fabianhjr) [reply pkgs/development/compilers/idris2/idris2.nix:94]: sure, no problem reviewing

Comment (fabianhjr) [reply pkgs/development/compilers/idris2/idris2.nix:94]: Merged after reviewing, thanks for contributing to nixpkgs! UwU

**Comment by fabianhjr** [comment]
Result of `nixpkgs-review pr 278391` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by mattpolzin** [comment]
Result of `nixpkgs-review pr 278391` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by mattpolzin** [comment]
Odd; where's my gratifying "1 package built" output. I'm reasonably sure my rebase didn't cause there to be no `idris2` package to build.

**Comment by fabianhjr** [comment]
probably detected as no change? or at least I would hope no change.

**Comment by mattpolzin** [comment]
Can confirm no change; building locally just downloads.

**Pull request merged**